lib/rules/restricted-component-names.js

```diff
@@ -0,0 +1,90 @@
+/**
+ * @author Wayne Zhang
+ * See LICENSE file in root directory for full license.
+ */
+'use strict'
+
+const utils = require('../utils')
+const { toRegExp } = require('../utils/regexp')
+
+const htmlElements = require('../utils/html-elements.json')
+const deprecatedHtmlElements = require('../utils/deprecated-html-elements.json')
+const svgElements = require('../utils/svg-elements.json')
+
+const RESERVED_NAMES_IN_VUE = new Set(
+  require('../utils/vue2-builtin-components')
+)
+const RESERVED_NAMES_IN_VUE3 = new Set(
+  require('../utils/vue3-builtin-components')
+)
+
+const RESERVED_NAMES_IN_HTML = new Set(htmlElements)
+const RESERVED_NAMES_IN_OTHERS = new Set([
+  ...deprecatedHtmlElements,
+  ...svgElements
+])
+
+const reservedNames = new Set([
+  ...RESERVED_NAMES_IN_HTML,
+  ...RESERVED_NAMES_IN_VUE,
+  ...RESERVED_NAMES_IN_VUE3,
+  ...RESERVED_NAMES_IN_OTHERS
+])
```

Comment:
There is no need to create multiple `Set`s from arrays, only to spread them into another array that is used to create the final `reservedNames` `Set`:

```suggestion
const htmlElements = require('../utils/html-elements.json')
const deprecatedHtmlElements = require('../utils/deprecated-html-elements.json')
const svgElements = require('../utils/svg-elements.json')
const vue2builtinComponents = require('../utils/vue2-builtin-components')
const vue3builtinComponents = require('../utils/vue3-builtin-components')

const reservedNames = new Set([
  ...htmlElements,
  ...deprecatedHtmlElements,
  ...svgElements,
  ...vue2builtinComponents,
  ...vue3builtinComponents
])
```